monai/csrc/filtering/trainable_bilateral/bf_layer_gpu_forward.cu

```diff
@@ -0,0 +1,313 @@
+    dO_dsig_y[batchOffset + homeOffset + c * cColorStride] =
+        -(1 / weightSum) * (valueSum / weightSum) * ySum_w + (1 / weightSum) * ySum_alpha;
+    dO_dsig_z[batchOffset + homeOffset + c * cColorStride] =
+        -(1 / weightSum) * (valueSum / weightSum) * zSum_w + (1 / weightSum) * zSum_alpha;
+  }
+}
+
+template <int C, int D>
+void BilateralFilterCudaForwardFunction(
+    torch::Tensor inputTensor,
+    torch::Tensor outputTensor,
+    torch::Tensor outputWeightsTensor,
+    torch::Tensor dO_dx_ki,
+    torch::Tensor dO_dsig_r,
+    torch::Tensor dO_dsig_x,
+    torch::Tensor dO_dsig_y,
+    torch::Tensor dO_dsig_z,
+    float sigma_x,
+    float sigma_y,
+    float sigma_z,
+    float colorSigma) {
+  // Getting tensor description.
+  TensorDescription desc = TensorDescription(inputTensor);
+
+  // Pre-calculating gaussian kernel.
+  int windowSize_x = std::max(((int)ceil(5.0f * sigma_x) | 1), 5); // ORing last bit to ensure odd window size
+  int windowSize_y = std::max(((int)ceil(5.0f * sigma_y) | 1), 5); // ORing last bit to ensure odd window size
+  int windowSize_z = std::max(((int)ceil(5.0f * sigma_z) | 1), 5); // ORing last bit to ensure odd window size
+  int halfWindowSize_x = floor(0.5f * windowSize_x);
+  int halfWindowSize_y = floor(0.5f * windowSize_y);
+  int halfWindowSize_z = floor(0.5f * windowSize_z);
+  int halfWindowSize_arr[] = {halfWindowSize_x, halfWindowSize_y, halfWindowSize_z};
+  float spatialExpConstant_x = -1.0f / (2 * sigma_x * sigma_x);
+  float spatialExpConstant_y = -1.0f / (2 * sigma_y * sigma_y);
+  float spatialExpConstant_z = -1.0f / (2 * sigma_z * sigma_z);
+  float colorExpConstant = -1.0f / (2 * colorSigma * colorSigma);
+
+  int* kernelSizes = new int[desc.dimensions];
+  kernelSizes[0] = windowSize_x;
+  kernelSizes[1] = windowSize_y;
+  kernelSizes[2] = windowSize_z;
+
+  auto* gaussianKernel_x = new float[windowSize_x];
+  auto* gaussianKernel_y = new float[windowSize_y];
+  auto* gaussianKernel_z = new float[windowSize_z];
+  auto* xDistanceSquared = new float[windowSize_x];
+  auto* yDistanceSquared = new float[windowSize_y];
+  auto* zDistanceSquared = new float[windowSize_z];
+
+  for (int i = 0; i < windowSize_x; i++) {
+    int distance = i - halfWindowSize_x;
+    gaussianKernel_x[i] = exp(distance * distance * spatialExpConstant_x);
+    xDistanceSquared[i] = distance * distance;
+  }
+  for (int i = 0; i < windowSize_y; i++) {
+    int distance = i - halfWindowSize_y;
+    gaussianKernel_y[i] = exp(distance * distance * spatialExpConstant_y);
+    yDistanceSquared[i] = distance * distance;
+  }
+  for (int i = 0; i < windowSize_z; i++) {
+    int distance = i - halfWindowSize_z;
+    gaussianKernel_z[i] = exp(distance * distance * spatialExpConstant_z);
+    zDistanceSquared[i] = distance * distance;
+  }
+
+  // Writing constant memory.
+  cudaMemcpyToSymbol(cBatchStride, &desc.batchStride, sizeof(int));
```

Comment:
Scaler and array with known (compiler known e.g. 3 here) length will be passed through constant cache (const memory) by default. So if you want to leverage constant cache here, you can pass desc.batchStride directly to the CUDA kernel by function argument.

Reply:
Thank you so much for your comments and having a look at the pull request, @zehuanw!
I am not sure if I understand you right here. Do you really think this would improve performance if we pass the desc.batchStride via the kernel arguments? Usually it should be more efficient to keep it in the constant cache for repeated kernel calls, right? Here, it should not make much of a difference anyways because of the small size of desc.batchStride, right? But please correct me if I am wrong.

Reply:
> Do you really think this would improve performance if we pass the desc.batchStride via the kernel arguments? 

It may not improve the performance. I meant that it will do no harm to your performance but just keep the code cleaner if you pass desc.batchStride directly to kernel with arguments. It's not a big problem anyway, so it's just FYI and please be free to adopt this idea or not.

>Usually it should be more efficient to keep it in the constant cache for repeated kernel calls, right?
Constant Cache is a on Chip cache (like shared memory) and you cannot keep something in it across different kernel launch (but L2 can). What you can do is copy data into Constant memory (on GPU memory) once and reuse in different kernels (cache will be cleared after kernel end). From the code it looks like you will call `cudaMemcpyToSymbol` for every forward and each forward has one kernel launch in this layer. So it looks like even we cannot save the memory copy here and reuse in different forwards with the same copy of constants？

Reply:
Okay, I see. Thank you for your explanation! You are right, there is only one kernel launch per forward/backward. I tried to use the shared memory the same way it was used in the [existing bilateral filter implementation](https://github.com/Project-MONAI/MONAI/blob/734d4ffc521adf3359b22b7a38c4eb7d3966c223/monai/csrc/filtering/bilateral/bilateralfilter_cuda.cu#216). If you are fine with it then I would keep the current implementation. But thank you again for pointing that out, I did not really think about the memory access here!